### PR TITLE
implement `reth p2p nodeset info`

### PR DESCRIPTION
**Pull request by int88**
```bash
# devp2p of geth
root@dev:~# devp2p nodeset info all-nodes.json 
Set contains 491 nodes.
ENR attribute counts:
       bsc: 6
       eth: 243
        id: 491
        ip: 491
       ip6: 1
     opera: 1
 secp256k1: 491
      snap: 475
       tcp: 490
       udp: 491

# reth
root@dev:~# reth p2p nodeset info --file all-nodes.json 
2024-07-06T10:51:09.419338Z  INFO Initialized tracing, debug log directory: /root/.cache/reth/logs/mainnet
Set contains 491 nodes
      bsc: 6
      eth: 243
       id: 491
       ip: 491
      ip6: 1
    opera: 1
secp256k1: 491
     snap: 475
      tcp: 490
      udp: 491
```

implement part of #7411

### CODE REVIEW

**Comment by int88** [comment]
@mattsse @emhane @Rjected PATL 🚀

File: bin/reth/src/commands/p2p/nodeset.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+impl Command {
+    /// Execute `p2p nodeset` command
+    pub fn execute(self) -> eyre::Result<()> {
+        match self.subcommand {
+            Subcommands::Info { file } => {
+                let mut f = File::open(file)?;
+                let mut content = String::new();
+                f.read_to_string(&mut content).expect("failed to read file");
```
Comment (mattsse) [inline bin/reth/src/commands/p2p/nodeset.rs:32]: this can be done in one call, but see also reth-fs-util for reading json file

**Review by mattsse — CHANGES_REQUESTED** [review]
supportive, but this should be moved to crates/cli/commands/

File: bin/reth/src/commands/nodeset.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+#[derive(Subcommand, Debug)]
+enum Subcommands {
+    /// Show statistics about a node set
+    Info {
+        /// The path of the JSON file used to store node set.
+        #[arg(long, value_name = "FILE", verbatim_doc_comment)]
+        file: PathBuf,
+    },
```
Comment (emhane) [inline bin/reth/src/commands/nodeset.rs:76]: let's use the `known-peers.json` file as default

File: bin/reth/src/commands/nodeset.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Record {
+    seq: u64,
+    record: Enr<SecretKey>,
+    score: i64,
+    firstResponse: DateTime<Utc>,
+    lastResponse: DateTime<Utc>,
+    lastCheck: DateTime<Utc>,
+}
```
Comment (emhane) [inline bin/reth/src/commands/nodeset.rs:26]: we don't use any of the fields in this type except `record`, so let's delete this type

please open an issue for writing data for the other fields to the `known-peers.json` file, and we can discuss whether we should do that in the issue

File: bin/reth/src/commands/nodeset.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#[derive(Subcommand, Debug)]
+enum Subcommands {
+    /// Show statistics about a node set
+    Info {
+        /// The path of the JSON file used to store node set.
+        #[arg(long, value_name = "FILE", verbatim_doc_comment)]
+        file: PathBuf,
+    },
+    /// Filters a node set
+    Filter,
```
Comment (emhane) [inline bin/reth/src/commands/nodeset.rs:78]: remove this if it's not implemented, and just add a TODO in a code comment, with a link to the specs and/or ref to geth implementation

**Review by emhane — CHANGES_REQUESTED** [review]
I love this. since the docs for "Node Set Utilities" aren't describing what the command should output, I guess it's up to the implementation. nonetheless, do you have any permalinks to the geth code used as ref?

**Comment by int88** [comment]
> I love this. since the docs for "Node Set Utilities" aren't describing what the command should output, I guess it's up to the implementation. nonetheless, do you have any permalinks to the geth code used as ref?

https://github.com/ethereum/go-ethereum/blob/master/cmd/devp2p/nodesetcmd.go

@emhane

File: bin/reth/src/commands/nodeset.rs (hunk 3 above)
Comment (int88) [reply bin/reth/src/commands/nodeset.rs:26]: These fields will be used when implement `nodeset filter`.

**Comment by int88** [comment]
And the content of node json file looks like:

```json
{
    "00051d84add764f6a1c154f83e50760579d644d24335a24ac23f4f46fd5da2e1": {
        "seq": 1711864314038,
        "record": "enr:-KO4QMne-2ViZIH_qe4P3D7DiXtk3WtEmyWvJ5-epTBPM8U5Nsy1msLffx-v2asiF9v0F7t8MKHjynz49-IQ6yhZfI6GAY6TEHi2g2V0aMfGhPCXvBOAgmlkgnY0gmlwhFf5iVmJc2VjcDI1NmsxoQNwGDtgE0IIvlWTJKr4duH8fWIfrk4NRCWk6uG9jmMILoRzbmFwwIN0Y3CCdl-DdWRwgndp",
        "score": 2,
        "firstResponse": "2024-06-15T07:19:53Z",
        "lastResponse": "2024-06-15T07:52:26Z",
        "lastCheck": "2024-06-15T07:52:26Z"
    },
    "002c83a14256a9eb64ab2bb49c242ff08e3ff49436bdde4936b465a2b7567cf4": {
        "seq": 1708643977995,
        "record": "enr:-KO4QOqf57C263tzY_lHLK537r9w_Kr1W-ecYL76GxGr1fFvA6M2rreakRX5TLBSdB9NUYtzAD8kw-18KKmk6H10LA2GAY3THgsLg2V0aMfGhPCXvBOAgmlkgnY0gmlwhE9_4h6Jc2VjcDI1NmsxoQPGfTTUwAW7M10Dg0zMNZ0CK2ILHbjB9f1fb4SBqlW1qYRzbmFwwIN0Y3CCdl-DdWRwgnZf",
        "score": 2,
        "firstResponse": "2024-06-15T07:20:03Z",
        "lastResponse": "2024-06-15T07:52:26Z",
        "lastCheck": "2024-06-15T07:52:26Z"
    },
    "009a50ae2ce73601eb29948f839c3f2997a692ab4192be031b7318036396d3a9": {
        "seq": 1681819827451,
        "record": "enr:-Ka4QL4wNNmamkysxv_xkEtSDTDqMUX32pxXnN2NYC8I2wCPDiVJEiJO0RrFBObWlmZaZrWzHUpKrmaYG7i2jA8RIRSGAYeURfz7g2JzY8CDZXRoy8qEe2_1C4Rmc8aMgmlkgnY0gmlwhLKipaqJc2VjcDI1NmsxoQNwbZ_dtQILf7mOsMFuSxlvPZ7CsN3sIjH_EK9vYHFpdoN0Y3CCdmeDdWRwgnZn",
        "score": 1,
        "firstResponse": "2024-06-15T07:52:51Z",
        "lastResponse": "2024-06-15T07:52:51Z",
        "lastCheck": "2024-06-15T07:52:51Z"
    },
...
}
```

**Comment by emhane** [comment]
> And the content of node json file looks like:
> 
> ```json
> {
>     "00051d84add764f6a1c154f83e50760579d644d24335a24ac23f4f46fd5da2e1": {
>         "seq": 1711864314038,
>         "record": "enr:-KO4QMne-2ViZIH_qe4P3D7DiXtk3WtEmyWvJ5-epTBPM8U5Nsy1msLffx-v2asiF9v0F7t8MKHjynz49-IQ6yhZfI6GAY6TEHi2g2V0aMfGhPCXvBOAgmlkgnY0gmlwhFf5iVmJc2VjcDI1NmsxoQNwGDtgE0IIvlWTJKr4duH8fWIfrk4NRCWk6uG9jmMILoRzbmFwwIN0Y3CCdl-DdWRwgndp",
>         "score": 2,
>         "firstResponse": "2024-06-15T07:19:53Z",
>         "lastResponse": "2024-06-15T07:52:26Z",
>         "lastCheck": "2024-06-15T07:52:26Z"
>     },
>     "002c83a14256a9eb64ab2bb49c242ff08e3ff49436bdde4936b465a2b7567cf4": {
>         "seq": 1708643977995,
>         "record": "enr:-KO4QOqf57C263tzY_lHLK537r9w_Kr1W-ecYL76GxGr1fFvA6M2rreakRX5TLBSdB9NUYtzAD8kw-18KKmk6H10LA2GAY3THgsLg2V0aMfGhPCXvBOAgmlkgnY0gmlwhE9_4h6Jc2VjcDI1NmsxoQPGfTTUwAW7M10Dg0zMNZ0CK2ILHbjB9f1fb4SBqlW1qYRzbmFwwIN0Y3CCdl-DdWRwgnZf",
>         "score": 2,
>         "firstResponse": "2024-06-15T07:20:03Z",
>         "lastResponse": "2024-06-15T07:52:26Z",
>         "lastCheck": "2024-06-15T07:52:26Z"
>     },
>     "009a50ae2ce73601eb29948f839c3f2997a692ab4192be031b7318036396d3a9": {
>         "seq": 1681819827451,
>         "record": "enr:-Ka4QL4wNNmamkysxv_xkEtSDTDqMUX32pxXnN2NYC8I2wCPDiVJEiJO0RrFBObWlmZaZrWzHUpKrmaYG7i2jA8RIRSGAYeURfz7g2JzY8CDZXRoy8qEe2_1C4Rmc8aMgmlkgnY0gmlwhLKipaqJc2VjcDI1NmsxoQNwbZ_dtQILf7mOsMFuSxlvPZ7CsN3sIjH_EK9vYHFpdoN0Y3CCdmeDdWRwgnZn",
>         "score": 1,
>         "firstResponse": "2024-06-15T07:52:51Z",
>         "lastResponse": "2024-06-15T07:52:51Z",
>         "lastCheck": "2024-06-15T07:52:51Z"
>     },
> ...
> }
> ```

ok, thx. somewhere this nodes file must get written. where is the geth code for writing this file? because the command doesn't appear to launch a p2p network itself, so the file must be written by the regular node command, or some other command dedicated specifically to this.

**Comment by int88** [comment]
> > And the content of node json file looks like:
> > ```json
> > {
> >     "00051d84add764f6a1c154f83e50760579d644d24335a24ac23f4f46fd5da2e1": {
> >         "seq": 1711864314038,
> >         "record": "enr:-KO4QMne-2ViZIH_qe4P3D7DiXtk3WtEmyWvJ5-epTBPM8U5Nsy1msLffx-v2asiF9v0F7t8MKHjynz49-IQ6yhZfI6GAY6TEHi2g2V0aMfGhPCXvBOAgmlkgnY0gmlwhFf5iVmJc2VjcDI1NmsxoQNwGDtgE0IIvlWTJKr4duH8fWIfrk4NRCWk6uG9jmMILoRzbmFwwIN0Y3CCdl-DdWRwgndp",
> >         "score": 2,
> >         "firstResponse": "2024-06-15T07:19:53Z",
> >         "lastResponse": "2024-06-15T07:52:26Z",
> >         "lastCheck": "2024-06-15T07:52:26Z"
> >     },
> >     "002c83a14256a9eb64ab2bb49c242ff08e3ff49436bdde4936b465a2b7567cf4": {
> >         "seq": 1708643977995,
> >         "record": "enr:-KO4QOqf57C263tzY_lHLK537r9w_Kr1W-ecYL76GxGr1fFvA6M2rreakRX5TLBSdB9NUYtzAD8kw-18KKmk6H10LA2GAY3THgsLg2V0aMfGhPCXvBOAgmlkgnY0gmlwhE9_4h6Jc2VjcDI1NmsxoQPGfTTUwAW7M10Dg0zMNZ0CK2ILHbjB9f1fb4SBqlW1qYRzbmFwwIN0Y3CCdl-DdWRwgnZf",
> >         "score": 2,
> >         "firstResponse": "2024-06-15T07:20:03Z",
> >         "lastResponse": "2024-06-15T07:52:26Z",
> >         "lastCheck": "2024-06-15T07:52:26Z"
> >     },
> >     "009a50ae2ce73601eb29948f839c3f2997a692ab4192be031b7318036396d3a9": {
> >         "seq": 1681819827451,
> >         "record": "enr:-Ka4QL4wNNmamkysxv_xkEtSDTDqMUX32pxXnN2NYC8I2wCPDiVJEiJO0RrFBObWlmZaZrWzHUpKrmaYG7i2jA8RIRSGAYeURfz7g2JzY8CDZXRoy8qEe2_1C4Rmc8aMgmlkgnY0gmlwhLKipaqJc2VjcDI1NmsxoQNwbZ_dtQILf7mOsMFuSxlvPZ7CsN3sIjH_EK9vYHFpdoN0Y3CCdmeDdWRwgnZn",
> >         "score": 1,
> >         "firstResponse": "2024-06-15T07:52:51Z",
> >         "lastResponse": "2024-06-15T07:52:51Z",
> >         "lastCheck": "2024-06-15T07:52:51Z"
> >     },
> > ...
> > }
> > ```
> 
> ok, thx. somewhere this nodes file must get written. where is the geth code for writing this file? because the command doesn't appear to launch a p2p network itself, so the file must be written by the regular node command, or some other command dedicated specifically to this.

I use command like `devp2p discv4 crawl -timeout 30s all-nodes.json`, ref: https://geth.ethereum.org/docs/developers/geth-developer/dns-discovery-setup. And `devp2p `

File: Cargo.toml [hunk 5]
```diff
@@ -515,6 +516,7 @@ proptest-arbitrary-interop = "0.1.0"
 secp256k1 = { version = "0.29", default-features = false, features = [
     "global-context",
     "recovery",
+    "serde",
```
Comment (mattsse) [inline Cargo.toml:519]: this should not be enabled on workspace level,
please undo

File: Cargo.toml [hunk 6]
```diff
@@ -432,6 +432,7 @@ auto_impl = "1"
 aquamarine = "0.5"
 bytes = "1.5"
 bitflags = "2.4"
+chrono = { version = "0.4.38", features = ["serde"] }
```
Comment (mattsse) [inline Cargo.toml:435]: same, please undo and instead enable in commands crate

File: crates/cli/commands/src/p2p/nodeset.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Record {
+    seq: u64,
+    record: Enr<SecretKey>,
+    score: i64,
+    firstResponse: DateTime<Utc>,
+    lastResponse: DateTime<Utc>,
+    lastCheck: DateTime<Utc>,
+}
```
Comment (mattsse) [inline crates/cli/commands/src/p2p/nodeset.rs:26]: let's make all of these pub

File: crates/cli/commands/src/p2p/nodeset.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+impl Command {
+    /// Execute `p2p nodeset` command
+    pub fn execute(self) -> eyre::Result<()> {
+        match self.subcommand {
+            Subcommands::Info { file } => {
+                let content = fs::read_to_string(file)?;
+
+                let nodes: HashMap<String, Record> =
+                    serde_json::from_str(&content).expect("failed to deserialize json");
```
Comment (mattsse) [inline crates/cli/commands/src/p2p/nodeset.rs:36]: this can simply use `?`

File: crates/cli/commands/src/p2p/nodeset.rs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+                let mut result: Vec<_> = keys.iter().collect();
+
+                result.sort_by(|a, b| a.0.cmp(b.0));
+
+                for (key, count) in result {
+                    let key_len = key.len();
+                    let padding = " ".repeat(max_key_len - key_len);
+                    println!("{}{}: {}", padding, key, count);
+                }
```
Comment (mattsse) [inline crates/cli/commands/src/p2p/nodeset.rs:58]: what does this do?

**Comment by emhane** [comment]
> I use command like `devp2p discv4 crawl -timeout 30s all-nodes.json`, ref: https://geth.ethereum.org/docs/developers/geth-developer/dns-discovery-setup.

ah ok, then this command doesn't make much sense here don't you think? since we aren't implementing the discv4 commands. I wasn't aware the node set commands worked on a file. I'd suggest you try to port these changes into a pr for https://github.com/agemanning/enr-cli, probably best to open an issue first, to see if the command is desired.

the rlpx protocol is implementated by reth. focus on those commands.

**Review by mattsse — CHANGES_REQUESTED** [review]
some nits and 1 q

hmm, we don't really want to rewrite all of the existing tools for devp2p, so I'm actually skeptical we should proceed with this pr because not very useful in isolation?

**Comment by int88** [comment]
sad 🤣

**Pull request closed without merging**